keyvi/include/keyvi/util/msgpack_util.h

```diff
@@ -38,69 +40,111 @@
 namespace keyvi {
 namespace util {
 
-class msgpack_buffer : public msgpack::sbuffer {
- public:
-  using msgpack::sbuffer::sbuffer;
-
-  void set_single_precision_float() { single_precision_float_ = true; }
-
-  bool get_single_precision_float() const { return single_precision_float_; }
+inline void JsonToMsgPack(const rapidjson::Value& value, msgpack::packer<msgpack::sbuffer>* msgpack_packer,
+                          bool single_precision_float) {
+  switch (value.GetType()) {
+    case rapidjson::kArrayType:
+      msgpack_packer->pack_array(value.Size());
+      for (auto& v : value.GetArray()) {
+        JsonToMsgPack(v, msgpack_packer, single_precision_float);
+      }
+      break;
+    case rapidjson::kObjectType:
+      msgpack_packer->pack_map(value.MemberCount());
+      for (auto& v : value.GetObject()) {
+        msgpack_packer->pack_str(v.name.GetStringLength());
+        msgpack_packer->pack_str_body(v.name.GetString(), v.name.GetStringLength());
+        JsonToMsgPack(v.value, msgpack_packer, single_precision_float);
+      }
+      break;
+    case rapidjson::kNumberType:
+      if (value.IsDouble()) {
+        if (single_precision_float) {
+          msgpack_packer->pack_float(value.GetFloat());
+        } else {
+          msgpack_packer->pack_double(value.GetDouble());
+        }
+      } else if (value.IsUint64()) {
+        msgpack_packer->pack_uint64(value.GetUint64());
+      } else {
+        msgpack_packer->pack_int64(value.GetInt64());
+      }
+      break;
+    case rapidjson::kStringType:
+      msgpack_packer->pack_str(value.GetStringLength());
+      msgpack_packer->pack_str_body(value.GetString(), value.GetStringLength());
+      break;
+    case rapidjson::kFalseType:
+      msgpack_packer->pack_false();
+      break;
+    case rapidjson::kTrueType:
+      msgpack_packer->pack_true();
+      break;
+    case rapidjson::kNullType:
+      msgpack_packer->pack_nil();
+      break;
+  }
+}
 
- private:
-  bool single_precision_float_ = false;
-};
+inline void MsgPackDump(rapidjson::Writer<rapidjson::StringBuffer>* writer, const msgpack::object& o) {
+  switch (o.type) {
+    case msgpack::type::NIL:
+      writer->Null();
+      break;
+
+    case msgpack::type::BOOLEAN:
+      writer->Bool(o.via.boolean);
+      break;
+
+    case msgpack::type::POSITIVE_INTEGER:
+      writer->Uint64(o.via.u64);
+      break;
+
+    case msgpack::type::NEGATIVE_INTEGER:
+      writer->Int64(o.via.i64);
+      break;
+
+    case msgpack::type::FLOAT:
+      writer->Double(o.via.f64);
+      break;
```

Comment:
In [JsonToMsgPack](https://github.com/KeyviDev/keyvi/pull/172/files#diff-9b644e267a5fc2eb412c40993bfd5740eb1cc692ca8801a1be8c5209dc15fc02R61-R66) we are writing floating point numbers calling 2 different methods: `pack_float()` and `pack_double()` and they have different type codes in `msgpack` format spec.

But in here while reading we have case just for `msgpack::type::FLOAT`, not sure if this is enough or we should have separate case for `msgpack::type::DOUBLE` as well.

Reply:
`msgpack::type::DOUBLE` is deprecated, we do not use it (and if, it's an alias for `FLOAT`). Encoding and decoding is not 1:1. That means when reading back there is no difference and json does not care either.

Reply:
FWIW I think I saw a distinction in newer versions for msgpack-c, will keep that in mind, once I update it, which is what I plan for after merging this PR.

Reply:
Cool, btw master is green now.